File: CS/ScrollPictureEdit/ScrollPictureEdit.cs
using System;
using System.ComponentModel;
using System.Drawing;
using System.Windows.Forms;
using DevExpress.XtraEditors;

namespace nsScrollPictureEdit
{
    class ScrollPictureEdit : PictureEdit
    {
        private DevExpress.XtraEditors.VScrollBar vScroll = null;
        private DevExpress.XtraEditors.HScrollBar hScroll = null;

        static ScrollPictureEdit()
        {
            RepositoryItemScrollPictureEdit.Register();
        }

        public ScrollPictureEdit()
            : base()
        {
            CreateVerticalScrollBar();
            CreateHorizontalScrollBar();

            hScroll.Show();
            vScroll.Show();
        }

        private void CreateVerticalScrollBar()
        {
            if ( vScroll != null )
                return;

            vScroll = new DevExpress.XtraEditors.VScrollBar();
            vScroll.Parent = this;
            vScroll.ScrollBarAutoSize = true;

            vScroll.Scroll += new ScrollEventHandler(OnVerticalScroll);
            vScroll.HandleCreated += new EventHandler(OnScrollBarHandleCreated);
        }

        private void CreateHorizontalScrollBar()
        {
            if ( hScroll != null )
                return;

            hScroll = new DevExpress.XtraEditors.HScrollBar();
            hScroll.Parent = this;
            hScroll.ScrollBarAutoSize = true;

            hScroll.Scroll += new ScrollEventHandler(OnHorizontalScroll);
            hScroll.HandleCreated += new EventHandler(OnScrollBarHandleCreated);
        }

        protected virtual void OnScrollBarHandleCreated(object sender, EventArgs e)
        {
            ((DevExpress.XtraEditors.ScrollBarBase)sender).Value = 0;
        }

        protected virtual void OnVerticalScroll(Object sender, ScrollEventArgs e)
        {
            this.Refresh();
        }

        protected virtual void OnHorizontalScroll(Object sender, ScrollEventArgs e)
        {
            this.Refresh();
        }

        protected virtual int CalcVScrollBarMaximum()
        {
            if ( this.Image.Width > this.Width )
                return this.Image.Height - this.Height + vScroll.LargeChange - 1 + hScroll.Height;

            return this.Image.Height - this.Height + vScroll.LargeChange - 1;
        }

        protected virtual int CalcHScrollBarMaximum()
        {
            if ( this.Image.Height > this.Height )
                return this.Image.Width - this.Width + hScroll.LargeChange - 1 + vScroll.Width;

            return this.Image.Width - this.Width + hScroll.LargeChange - 1;
        }

        protected virtual void UpdateScrollBars()
        {
            if ( this.Image != null && this.Properties.Scrollable )
            {
                if ( this.Image.Height > this.Height )
                {
                    vScroll.Left = this.Width - vScroll.Width;
                    vScroll.Height = this.Image.Width > this.Width ? this.Height - hScroll.Height : this.Height;
                    vScroll.Maximum = CalcVScrollBarMaximum();
                    vScroll.Show();
                }
                else
                    vScroll.Hide();

                if ( this.Image.Width > this.Width )
                {
                    hScroll.Top = this.Height - hScroll.Height;
                    hScroll.Width = this.Image.Height > this.Height ? this.Width - vScroll.Width : this.Width;
                    hScroll.Maximum = CalcHScrollBarMaximum();
                    hScroll.Show();
                }
                else
                    hScroll.Hide();
            }
            else
            {
                vScroll.Hide();
                hScroll.Hide();
            }
        }

        protected override void LayoutChanged()
        {
            base.LayoutChanged();
            UpdateScrollBars();
        }

        protected override void OnPaint(PaintEventArgs e)
        {
            if ( this.Image != null && this.Properties.Scrollable )
            {
                e.Graphics.DrawImage(this.Image, e.ClipRectangle, hScroll.Value, vScroll.Value, e.ClipRectangle.Width, e.ClipRectangle.Height, GraphicsUnit.Pixel);

                if ( vScroll.Visible && hScroll.Visible )
                {
                    Rectangle blackRect = new Rectangle(hScroll.Width, hScroll.Top, vScroll.Left + vScroll.Width, hScroll.Top + hScroll.Height);
                    e.Graphics.FillRectangle(new SolidBrush(this.BackColor), blackRect);
                }
            }
            else
                base.OnPaint(e);
        }

        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public new RepositoryItemScrollPictureEdit Properties
        {
            get { return base.Properties as RepositoryItemScrollPictureEdit; }
        }

        public override string EditorTypeName
        {
            get { return RepositoryItemScrollPictureEdit.EditorName; }
        }
    }
}

File: VB/ScrollPictureEdit/ScrollPictureEdit.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.ComponentModel
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.XtraEditors

Namespace nsScrollPictureEdit
	Friend Class ScrollPictureEdit
		Inherits PictureEdit
		Private vScroll As DevExpress.XtraEditors.VScrollBar = Nothing
		Private hScroll As DevExpress.XtraEditors.HScrollBar = Nothing

		Shared Sub New()
			RepositoryItemScrollPictureEdit.Register()
		End Sub

		Public Sub New()
			MyBase.New()
			CreateVerticalScrollBar()
			CreateHorizontalScrollBar()

			hScroll.Show()
			vScroll.Show()
		End Sub

		Private Sub CreateVerticalScrollBar()
			If vScroll IsNot Nothing Then
				Return
			End If

			vScroll = New DevExpress.XtraEditors.VScrollBar()
			vScroll.Parent = Me
			vScroll.ScrollBarAutoSize = True

			AddHandler vScroll.Scroll, AddressOf OnVerticalScroll
			AddHandler vScroll.HandleCreated, AddressOf OnScrollBarHandleCreated
		End Sub

		Private Sub CreateHorizontalScrollBar()
			If hScroll IsNot Nothing Then
				Return
			End If

			hScroll = New DevExpress.XtraEditors.HScrollBar()
			hScroll.Parent = Me
			hScroll.ScrollBarAutoSize = True

			AddHandler hScroll.Scroll, AddressOf OnHorizontalScroll
			AddHandler hScroll.HandleCreated, AddressOf OnScrollBarHandleCreated
		End Sub

		Protected Overridable Sub OnScrollBarHandleCreated(ByVal sender As Object, ByVal e As EventArgs)
			CType(sender, DevExpress.XtraEditors.ScrollBarBase).Value = 0
		End Sub

		Protected Overridable Sub OnVerticalScroll(ByVal sender As Object, ByVal e As ScrollEventArgs)
			Me.Refresh()
		End Sub

		Protected Overridable Sub OnHorizontalScroll(ByVal sender As Object, ByVal e As ScrollEventArgs)
			Me.Refresh()
		End Sub

		Protected Overridable Function CalcVScrollBarMaximum() As Integer
			If Me.Image.Width > Me.Width Then
				Return Me.Image.Height - Me.Height + vScroll.LargeChange - 1 + hScroll.Height
			End If

			Return Me.Image.Height - Me.Height + vScroll.LargeChange - 1
		End Function

		Protected Overridable Function CalcHScrollBarMaximum() As Integer
			If Me.Image.Height > Me.Height Then
				Return Me.Image.Width - Me.Width + hScroll.LargeChange - 1 + vScroll.Width
			End If

			Return Me.Image.Width - Me.Width + hScroll.LargeChange - 1
		End Function

		Protected Overridable Sub UpdateScrollBars()
			If Me.Image IsNot Nothing AndAlso Me.Properties.Scrollable Then
				If Me.Image.Height > Me.Height Then
					vScroll.Left = Me.Width - vScroll.Width
					If Me.Image.Width > Me.Width Then
						vScroll.Height = Me.Height - hScroll.Height
					Else
						vScroll.Height = Me.Height
					End If
					vScroll.Maximum = CalcVScrollBarMaximum()
					vScroll.Show()
				Else
					vScroll.Hide()
				End If

				If Me.Image.Width > Me.Width Then
					hScroll.Top = Me.Height - hScroll.Height
					If Me.Image.Height > Me.Height Then
						hScroll.Width = Me.Width - vScroll.Width
					Else
						hScroll.Width = Me.Width
					End If
					hScroll.Maximum = CalcHScrollBarMaximum()
					hScroll.Show()
				Else
					hScroll.Hide()
				End If
			Else
				vScroll.Hide()
				hScroll.Hide()
			End If
		End Sub

		Protected Overrides Sub LayoutChanged()
			MyBase.LayoutChanged()
			UpdateScrollBars()
		End Sub

		Protected Overrides Sub OnPaint(ByVal e As PaintEventArgs)
			If Me.Image IsNot Nothing AndAlso Me.Properties.Scrollable Then
				e.Graphics.DrawImage(Me.Image, e.ClipRectangle, hScroll.Value, vScroll.Value, e.ClipRectangle.Width, e.ClipRectangle.Height, GraphicsUnit.Pixel)

				If vScroll.Visible AndAlso hScroll.Visible Then
					Dim blackRect As New Rectangle(hScroll.Width, hScroll.Top, vScroll.Left + vScroll.Width, hScroll.Top + hScroll.Height)
					e.Graphics.FillRectangle(New SolidBrush(Me.BackColor), blackRect)
				End If
			Else
				MyBase.OnPaint(e)
			End If
		End Sub

		<DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
		Public Shadows ReadOnly Property Properties() As RepositoryItemScrollPictureEdit
			Get
				Return TryCast(MyBase.Properties, RepositoryItemScrollPictureEdit)
			End Get
		End Property

		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return RepositoryItemScrollPictureEdit.EditorName
			End Get
		End Property
	End Class
End Namespace